botocove/cove_host_account.py

```diff
@@ -102,28 +103,107 @@ def _get_boto3_sts_client(self, assuming_session: Optional[Session]) -> STSClien
         else:
-            validated_ignore_ids = set()
+            ignore_accounts_set = set()
 
         if target_ids is None:
             # No target_ids passed
-            target_accounts = self._gather_org_assume_targets()
+            target_accounts_set = self._gather_org_assume_targets()
         else:
-            # Specific list of IDs passed
-            target_accounts = set(target_ids)
-
-        return target_accounts - validated_ignore_ids
-
-    def _format_ignore_ids(self) -> Set[str]:
-        if not isinstance(self.provided_ignore_ids, list):
-            raise TypeError("ignore_ids must be a list of account IDs")
-        for account_id in self.provided_ignore_ids:
-            if len(account_id) != 12:
-                raise TypeError("All ignore_id in list must be 12 character strings")
-            if not isinstance(account_id, str):
-                raise TypeError("All ignore_id list entries must be strings")
-        return set(self.provided_ignore_ids)
+            target_accounts, target_ous = self._get_validated_ids(
+                target_ids, "target_ids"
+            )
+            target_accounts_from_ous = self._get_all_accounts_by_organization_units(
+                target_ous
+            )
+            target_accounts.extend(target_accounts_from_ous)
+            target_accounts_set = set(target_accounts)
+
+        return target_accounts_set - ignore_accounts_set
+
+    def _get_validated_ids(
+        self, ids: List[str], ids_type: str
+    ) -> Tuple[List[str], List[str]]:
+
+        accounts: List[str] = []
+        ous: List[str] = []
+
+        for current_id in ids:
+            if not isinstance(current_id, str):
+                raise TypeError("All {ids_type} must be strings")
+            if re.match(r"^\d{12}$", current_id):
+                accounts.append(current_id)
+                continue
+            if re.match(r"^ou-[0-9a-z]{4,32}-[a-z0-9]{8,32}$", current_id):
+                ous.append(current_id)
+                continue
+            raise ValueError(
+                f"{ids_type} entry is neither an aws account nor an ou: {current_id}"
+            )
+
+        return accounts, ous
+
+    def _get_all_accounts_by_organization_units(
+        self, target_ous: List[str]
+    ) -> List[str]:
+
+        account_list: List[str] = []
+
+        for parent_ou in target_ous:
+
+            current_ou_list: List[str] = []
+
+            # current_ou_list is mutated and recursivly populated with all childs
+            self._get_all_child_ous(parent_ou, current_ou_list)
+
+            # for complete list add parent ou as well to list of child ous
+            current_ou_list.append(parent_ou)
+
+            account_list.extend(
+                self._get_accounts_by_organization_units(current_ou_list)
+            )
+
+        return account_list
+
+    def _get_all_child_ous(self, parent_ou: str, ou_list: List[str]) -> None:
+
+        child_ous = (
```

Comment:
let's break both paginators out into function calls so we can use a cache (functools.lru_cache would be my suggestion) to avoid re-doing work when traversing now that we're potentially going to be walking the same OU tree twice or more

(for example - in OU A, containing B and C, we want to pass A as a target but ignore C - we'd look up children of C twice, and contents of C twice without caching)

Reply:
Great, thanks for your valuable feedback! And lets patch this optimisation later